tasking/pulp/tasking/base.py

```diff
@@ -0,0 +1,444 @@
+        try:
+            worker = ReservedResource.objects.get(resource=resource_id).worker
+        except ReservedResource.DoesNotExist:
+            pass
+        else:
+            break
+
+        try:
+            worker = _get_unreserved_worker()
+        except Worker.DoesNotExist:
+            pass
+        else:
+            break
+
+        # No worker is ready for this work, so we need to wait
+        time.sleep(0.25)
+
+    task_status = TaskStatus.objects.get(pk=task_id)
+    ReservedResource.objects.create(task=task_status, worker=worker, resource=resource_id)
+
+    inner_kwargs['routing_key'] = worker.name
+    inner_kwargs['exchange'] = DEDICATED_QUEUE_EXCHANGE
+    inner_kwargs['task_id'] = task_id
+
+    try:
+        celery.tasks[name].apply_async(*inner_args, **inner_kwargs)
+    finally:
+        _release_resource.apply_async((task_id, ), routing_key=worker.name,
+                                      exchange=DEDICATED_QUEUE_EXCHANGE)
+
+
+def _get_unreserved_worker():
+    """
+    Find an unreserved Worker
+
+    Return the Worker instance that has no ReservedResource associated with it. If all workers have
+    ReservedResource relationships a Worker.DoesNotExist pulp.app.model.Worker exception is
+    raised.
+
+    This function also provides randomization for worker selection.
+
+    :raises Worker.DoesNotExist: If all workers have ReservedResource entries associated with them.
+
+    :returns:          The Worker instance that has no reserved_resource
+                       entries associated with it.
+    :rtype:            pulp.app.model.Worker
+    """
+    free_workers = [Worker.objects.annotate(Count('reservations')).filter(reservations__count=0)]
+    if len(free_workers) == 0:
+        raise Worker.DoesNotExist()
+    random.shuffle(free_workers)
+    return free_workers[0]
+
+
+def delete_worker(name, normal_shutdown=False):
+    """
+    Delete the Worker with name from the database, cancel any associated tasks and reservations
+
+    If the worker shutdown normally, no message is logged, otherwise an error level message is
+    logged. Default is to assume the worker did not shut down normally.
+
+    Any resource reservations associated with this worker are cleaned up by this function.
+
+    Any tasks associated with this worker are explicitly canceled.
+
+    :param name:            The name of the worker you wish to delete.
+    :type  name:            basestring
+    :param normal_shutdown: True if the worker shutdown normally, False otherwise. Defaults to
+                            False.
+    :type normal_shutdown:  bool
+    """
+    if not normal_shutdown:
+        msg = _('The worker named %(name)s is missing. Canceling the tasks in its queue.')
+        msg = msg % {'name': name}
+        _logger.error(msg)
+
+    # Cancel all of the tasks that were assigned to this worker's queue
+    for task_status in Worker.objects.tasks.filter(state__in=TASK_INCOMPLETE_STATES):
```

Comment:
Doesn't there need to be a ``get(name=name)`` in here somewhere to scope it to one worker?

For that matter, it probably makes sense to get the worker object first, then use that object to both find tasks and later call its ``delete()`` method.